Add ContainerInfo for TaskSpecification

activatedgeek opened the pull request:
The existing `DefaultTaskSpecification` does not provide appropriate entry points for running containers as tasks. This PR proposes an updated `TaskSpecification` interface.

cc @gabrielhartmann

activatedgeek commented:
I would request somebody to point me to the final `TaskInfo` construction so that the `getContainer` interface method can be utilized to add `ContainerInfo` to the protobuf.

gabrielhartmann commented:
@activatedgeek:  Thanks for the PR.  Looks reasonable to me.  Could you add a few unit tests validating that this generates expected protobufs.  In particular I'm interested in validating that `getExistingOfferRequirement` works as expected after launching with a `ContainerInfo`.  Should be fine, but would be nice to validate with a test or two.

mohitsoni commented:
What happens when ContainerInfo and CommandInfo are set simultaneously ?

gabrielhartmann commented:
Also, @activatedgeek and @mohitsoni.  I this situation we'd probably want to make CommandInfo optional.  I think with entry points all combinations are possible, and I don't see how we could validate TaskSpecifications actually.  Maybe trust the developer?  A revolutionary thought.

activatedgeek commented:
@mohitsoni, From what I remember after seeing `mesos.proto`, I think `ContainerInfo` uses `CommandInfo` if it is available for the `[command]` part in `docker run [options] image [command]`). Please correct me if I am wrong.

@gabrielhartmann I did think about making both the `ContainerInfo` and `CommandInfo` optional which in fact does keep us in line with the protobuf definition anyways (where both are optional and "trust the developer" kind of thing). I believe it should be done and an exception can be thrown in case both are `empty()`. What do you think?

gabrielhartmann commented:
@activatedgeek:  I think we're getting somewhere.  I was trying to think what validation we could perform.  Both empty throwing the exception you specify sounds good to me.  Can't validate the developer hasn't done something incorrect if both are specified.

Making the change discussed above and writing some unit tests should be sufficient to get this across the line.

/cc @mohitsoni

activatedgeek commented:
@gabrielhartmann I rebased from master and updated my code. Requesting for review.

* To ensure that the developer doesn't end up missing both `CommandInfo` and `ContainerInfo`, only appropriate constructors have been provided. More precisely, constructor arguments of `CommandInfo` and `ContainerInfo` are not optional in the `DefaultTaskSpecification`.

* The static `DefaultTaskSpecification.create` method throws appropriate exceptions and uses appropriate constructors based on which of the two protobufs are available.

* Updated tests for `CustomTaskSet` (covering all possible constructors)

gabrielhartmann commented:
@mesosphere-ci: retest this please

gabrielhartmann commented on src/main/java/org/apache/mesos/offer/DefaultOfferRequirementProvider.java:
```diff
@@ -29,8 +29,9 @@ public DefaultOfferRequirementProvider(TaskConfigRouter taskConfigRouter) {
+
         Protos.CommandInfo updatedCommand = taskConfigRouter.getConfig(taskType)
-                .updateEnvironment(taskSpecification.getCommand());
+                .updateEnvironment(taskSpecification.getCommand().get());
```
Doesn't this need an `isPresent()` check before use?  If it's optional now we'll we want to skip adding it when it's not present correct?

activatedgeek replied:
Oh! That's a big miss. I also want to understand that why is there a special update method for the `CommandInfo`. Why not simply set it into the new `Protos.TaskInfo.Builder`?

gabrielhartmann commented on src/main/java/org/apache/mesos/specification/DefaultTaskSet.java:
```diff
@@ -16,6 +16,7 @@
     private final String name;
     private final List<TaskSpecification> taskSpecifications;
 
+    // TODO: this should be deprecated
```
Why should this be deprecated?  Creation with just a CommandInfo is valid.

activatedgeek replied:
Indeed that is valid but it misses out on various other optional elements and I thought it might be unnecessary to strip down the `TaskInfo` so much. But I guess, let's keep it.

gabrielhartmann commented:
@activatedgeek:  Looks good.  A couple comments.  Also, this needs OfferRequirement tests.  I'd like to see something that rejects creation of an OfferRequirement if input is bad, and validates expected content in expected cases.

gabrielhartmann commented on src/main/java/org/apache/mesos/specification/DefaultTaskSet.java:
```diff
@@ -58,6 +59,58 @@ public static DefaultTaskSet create(
+            Optional<PlacementRuleGenerator> placementOptional,
+            Optional<Protos.HealthCheck> healthCheck) {
+
+        List<TaskSpecification> taskSpecifications = new ArrayList<>();
```
A lot of code duplication in this loop and the loop below.

activatedgeek replied:
Here again, a private static common function should fix bad duplication.

activatedgeek replied:
Same as below, fixed via a `private static` sink function.

activatedgeek commented on src/main/java/org/apache/mesos/specification/DefaultTaskSpecification.java:
```diff
@@ -45,26 +46,102 @@ public static DefaultTaskSpecification create(Protos.TaskInfo taskInfo, Optional
+     * Create task with only {@link Protos.CommandInfo}.
+     */
+    protected DefaultTaskSpecification(
+            String name,
```
I'll fix this by making all the constructors now go via a private constructor. That removes the bad duplication. Thanks!

gabrielhartmann commented on src/main/java/org/apache/mesos/specification/DefaultTaskSet.java:
```diff
@@ -16,22 +16,23 @@
                 count,
                 name,
-                command,
+                Optional.empty(),
```
Please put comments back in for what the empty elements represent.

gabrielhartmann commented on src/main/java/org/apache/mesos/specification/DefaultTaskSet.java:
```diff
@@ -44,19 +45,16 @@ public static DefaultTaskSet create(
-        }
-
-        return create(name, taskSpecifications);
+        return create(
```
same here.

gabrielhartmann commented on src/main/java/org/apache/mesos/specification/DefaultTaskSet.java:
```diff
@@ -69,19 +67,16 @@ public static DefaultTaskSet create(
-        }
-
-        return create(name, taskSpecifications);
+        return create(
```
and here

gabrielhartmann commented:
@activatedgeek:  mostly copy/paste from Slack:

The test is fine but more is needed.  It looks like all the tests in `DefaultOfferRequirementProviderTest` are focused around testing already launched Tasks.  Ignoring that oddness, it would be nice to have a couple tests that use the `getNewOfferRequirement` method of the `DefaultOfferRequirementProvider` with the new kinds of `TaskSpecification` you’ve enabled.

And then check that the OfferRequirement has been constructed as expected.  One with just a ContainerInfo, one with just a CommandInfo, one with both, and one with neither should be sufficient.  Then there should be a couple tests that test the `getExistingOfferRequirement` method to make sure one can recover from crash and detect changes to the TaskSpec.

e.g. One might start with a valid TaskSpecification, and then change it so that it's no invalid.  That should cause a failure in generating an an `OfferRequirement`.

activatedgeek commented:
@gabrielhartmann Right, that falls in line. Thanks!

activatedgeek commented on sdk/core/src/main/java/org/apache/mesos/specification/DefaultService.java:
```diff
@@ -82,7 +82,7 @@ public void run() {
     private static void registerFramework(Scheduler sched, Protos.FrameworkInfo frameworkInfo, String masterUri) {
-        LOGGER.info("Registering framework: {}", frameworkInfo);
+        LOGGER.info(String.format("Registering framework: %s", frameworkInfo));
         SchedulerDriver driver = new SchedulerDriverFactory().create(sched, frameworkInfo, masterUri);
```
Oh I believe that happened after I had committed. I kept my changes during the rebase. I'll revert. Thanks!

gabrielhartmann commented:
@mesosphere-ci: retest this please
